File: examples/object_detection_yoloV7/run_object_detection.cpp
/**
 * @file run_object_detection.cpp
 * @brief Run a YOLOv7 image through CModelFacade and generic task adapters.
 */

#include <inference/model_facade.h>
#include <inference/task_adapters.h>
#include <utils/logging/CLogger.h>

#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <tclap/CmdLine.h>

#include <cmath>
#include <cstddef>
#include <cstdint>
#include <exception>
#include <filesystem>
#include <iostream>
#include <limits>
#include <optional>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace
{
    namespace fs = std::filesystem;
    namespace infer = ptafdeploy::inference;
    namespace logging = ptafdeploy::logging;

    [[nodiscard]] logging::CLogger& GetLogger()
    {
        static logging::CLogger logger("object_detection_yoloV7", logging::ELogLevel::Info,
                                       logging::ELogColorMode::Disabled, std::clog, std::clog);
        static const bool environment_applied = logger.setLevelFromEnvironment();
        static_cast<void>(environment_applied);
        return logger;
    }

    struct SArguments
    {
        fs::path manifest_path{};
        fs::path image_path{};
        std::optional<infer::EExecutionTarget> execution_target{};
        int device_id{0};
        float score_threshold{0.25F};
        size_t max_detections{20U};
    };

    [[nodiscard]] std::optional<infer::EExecutionTarget> ParseTarget(const std::string& value)
    {
        if (value == "manifest")
        {
            return std::nullopt;
        }
        if (value == "cpu")
        {
            return infer::EExecutionTarget::cpu;
        }
        if (value == "cuda")
        {
            return infer::EExecutionTarget::cuda;
        }

        throw std::invalid_argument("--target expects manifest, cpu, or cuda.");
    }

    [[nodiscard]] SArguments ParseArguments(const int argc, char** argv)
    {
        TCLAP::CmdLine command(
            "Run one image through CModelFacade, shared HWC-to-NCHW preprocessing, and a "
            "schema-driven pre-NMS detection decoder.",
            ' ', PTAFDEPLOY_CLI_VERSION);
        command.setExceptionHandling(false);

        TCLAP::UnlabeledValueArg<std::string> manifest_path(
            "model", "Path to the object-detection model manifest", true, "",
            "model.ptafmodel", command);
        TCLAP::UnlabeledValueArg<std::string> image_path(
            "image", "Path to the input image", true, "", "image", command);
        TCLAP::ValueArg<std::string> target(
            "", "target", "Execution target override or manifest-configured runtime", false,
            "manifest", "manifest|cpu|cuda", command);
        TCLAP::ValueArg<int> device(
            "", "device", "Non-negative runtime device index", false, 0,
            "non-negative integer", command);
        TCLAP::ValueArg<float> score_threshold(
            "", "score-threshold", "Minimum finite non-negative detection score", false, 0.25F,
            "non-negative number", command);
        TCLAP::ValueArg<long long> max_detections(
            "", "max-detections", "Maximum number of score-sorted detections to report", false,
            20LL, "non-negative integer", command);

        command.parse(argc, argv);

        if (device.getValue() < 0)
        {
            throw std::invalid_argument("--device expects a non-negative integer.");
        }
        if (!std::isfinite(score_threshold.getValue()) || score_threshold.getValue() < 0.0F)
        {
            throw std::invalid_argument("--score-threshold expects a finite non-negative number.");
        }
        if (max_detections.getValue() < 0 ||
            static_cast<unsigned long long>(max_detections.getValue()) >
                std::numeric_limits<size_t>::max())
        {
            throw std::invalid_argument("--max-detections expects a non-negative integer.");
        }

        SArguments arguments;
        arguments.manifest_path = manifest_path.getValue();
        arguments.image_path = image_path.getValue();
        arguments.execution_target = ParseTarget(target.getValue());
        arguments.device_id = device.getValue();
        arguments.score_threshold = score_threshold.getValue();
        arguments.max_detections = static_cast<size_t>(max_detections.getValue());

        return arguments;
    }

    void LoadModel(const SArguments& arguments, infer::CModelFacade& model)
    {
        if (!arguments.execution_target.has_value())
        {
            model.LoadModelConfig(arguments.manifest_path.string());
            return;
        }

        infer::SRuntimeConfig runtime;
        runtime.SetDeviceId(arguments.device_id);
        runtime.ClearExecutionTargetPriority();
        runtime.AddExecutionTarget(*arguments.execution_target);
        runtime.SetAllowFallback(false);
        model.LoadModelConfigWithRuntimeConfig(arguments.manifest_path.string(), runtime);
    }

    [[nodiscard]] infer::SFloatTensor PrepareImage(const cv::Mat& image,
                                                   const infer::STensorInfo& input_info)
    {
        if (input_info.dtype != "float32" || input_info.shape.size() != 4U ||
            input_info.shape[0] != 1 || input_info.shape[1] != 3 || input_info.shape[2] <= 0 ||
            input_info.shape[3] <= 0 || input_info.shape[2] > std::numeric_limits<int>::max() ||
            input_info.shape[3] > std::numeric_limits<int>::max())
        {
            throw std::runtime_error(
                "YOLO demo expects one float32 input with concrete shape [1,3,H,W].");
        }

        const auto height = static_cast<size_t>(input_info.shape[2]);
        const auto width = static_cast<size_t>(input_info.shape[3]);
        cv::Mat resized_image;
        cv::resize(image, resized_image,
                   cv::Size{static_cast<int>(width), static_cast<int>(height)}, 0.0, 0.0,
                   cv::INTER_LINEAR);
        if (!resized_image.isContinuous())
        {
            resized_image = resized_image.clone();
        }

        const auto* pixels = resized_image.ptr<uint8_t>();
        return infer::MakeNchwFloatTensorFromHwcAccessor(
            input_info.name, height, width, 3U, 1.0F / 255.0F, true,
            [pixels](const size_t index) { return pixels[index]; });
    }

    [[nodiscard]] std::string FormatShape(const std::vector<int64_t>& shape)
    {
        std::ostringstream stream;
        stream << '[';
        for (size_t index = 0; index < shape.size(); ++index)
        {
            if (index != 0U)
            {
                stream << ',';
            }
            stream << shape[index];
        }
        stream << ']';
        return stream.str();
    }

    void PrintDetections(const infer::SFloatTensor& output,
                         const std::vector<infer::SDetection2D>& detections)
    {
        std::cout << "output_name=" << output.name << '\n'
                  << "output_shape=" << FormatShape(output.shape) << '\n'
                  << "detections=" << detections.size() << '\n';
        for (size_t index = 0; index < detections.size(); ++index)
        {
            const infer::SDetection2D& detection = detections[index];
            std::cout << "detection[" << index
                      << "]=" << "class=" << detection.classification.class_id
                      << ",score=" << detection.classification.score
                      << ",cx=" << detection.bounds.center.x << ",cy=" << detection.bounds.center.y
                      << ",w=" << detection.bounds.size.width
                      << ",h=" << detection.bounds.size.height << '\n';
        }
    }
} // namespace

/**
 * @brief Run one image through the facade-backed YOLO integration.
 * @param argc Number of command-line arguments.
 * @param argv Command-line argument values.
 * @return Zero on success, one for invalid input or inference failure.
 */
int main(const int argc, char** argv)
{
    try
    {
        const SArguments arguments = ParseArguments(argc, argv);
        GetLogger().info("Loading YOLO model configuration: ", arguments.manifest_path.string());
        GetLogger().debug("Input image=", arguments.image_path.string(),
                          ";score_threshold=", arguments.score_threshold,
                          ";max_detections=", arguments.max_detections);
        infer::CModelFacade model;
        LoadModel(arguments, model);
        if (model.GetRole() != "object_detection")
        {
            throw std::runtime_error("YOLO demo requires role=object_detection in the manifest.");
        }
        if (model.GetNumInputs() != 1U || model.GetNumOutputs() != 1U)
        {
            throw std::runtime_error(
                "YOLO demo currently requires one model input and one output.");
        }

        const cv::Mat image = cv::imread(arguments.image_path.string(), cv::IMREAD_COLOR);
        if (image.empty())
        {
            throw std::invalid_argument("Could not read input image: " +
                                        arguments.image_path.string());
        }

        const infer::SFloatTensor input = PrepareImage(image, model.GetInputInfo(0));
        const infer::SFloatTensor output = model.InferSingleFloatTensor(input);
        if (output.shape.empty() || output.shape.back() < 6)
        {
            throw std::runtime_error(
                "YOLO demo expects an attribute axis containing xywh, objectness, and classes.");
        }

        infer::SDetectionRowSchema detection_schema;
        detection_schema.box_encoding = infer::EBoundingBoxEncoding::center_xywh;
        detection_schema.box_coordinate_0_index = 0U;
        detection_schema.box_coordinate_1_index = 1U;
        detection_schema.box_coordinate_2_index = 2U;
        detection_schema.box_coordinate_3_index = 3U;
        detection_schema.objectness_index = 4;
        detection_schema.first_class_score_index = 5U;
        detection_schema.class_score_count = static_cast<size_t>(output.shape.back() - 5);
        const std::vector<infer::SDetection2D> detections = infer::DecodeDetectionRows(
            output, detection_schema, arguments.score_threshold, arguments.max_detections);

        std::cout << "role=" << model.GetRole() << '\n'
                  << "backend=" << model.GetBackendDetail() << '\n'
                  << "input_name=" << input.name << '\n'
                  << "input_shape=" << FormatShape(input.shape) << '\n';
        PrintDetections(output, detections);
        return 0;
    }
    catch (const TCLAP::ExitException& exit_request)
    {
        return exit_request.getExitStatus();
    }
    catch (const TCLAP::ArgException& error)
    {
        GetLogger().error("Invalid command line: ", error.error(), " for ", error.argId());
        return 1;
    }
    catch (const std::exception& error)
    {
        GetLogger().error(error.what());
        return 1;
    }
}

File: examples/object_detection_yoloV7/run_object_detection.py
"""Run a YOLOv7 image through the generated PTAF Deploy Python wrapper."""

from __future__ import annotations

import argparse
from dataclasses import dataclass
from pathlib import Path
import sys
from typing import Sequence

import numpy as np
from PIL import Image

import autoforge_deploy as ptaf


@dataclass(frozen=True)
class DemoOptions:
    """Validated command-line options for one YOLO inference."""

    manifest_path: Path
    image_path: Path
    target: str
    device_id: int
    score_threshold: float
    max_detections: int


def parse_arguments(arguments: Sequence[str] | None = None) -> DemoOptions:
    """Parse command-line arguments.

    Args:
        arguments: Optional argument sequence. Defaults to ``sys.argv``.

    Returns:
        Validated demo options.
    """
    parser = argparse.ArgumentParser(
        description=(
            "Run one image through CModelFacade and a generic row decoder. "
            "Results are pre-NMS and no annotated image is rendered."
        )
    )
    parser.add_argument("manifest_path", type=Path)
    parser.add_argument("image_path", type=Path)
    parser.add_argument(
        "--target",
        choices=("manifest", "cpu", "cuda"),
        default="manifest",
        help="Use manifest runtime settings or force one execution target.",
    )
    parser.add_argument("--device", type=int, default=0, dest="device_id")
    parser.add_argument("--score-threshold", type=float, default=0.25)
    parser.add_argument("--max-detections", type=int, default=20)
    parsed = parser.parse_args(arguments)

    if parsed.device_id < 0:
        parser.error("--device must be non-negative")
    if not np.isfinite(parsed.score_threshold) or parsed.score_threshold < 0.0:
        parser.error("--score-threshold must be finite and non-negative")
    if parsed.max_detections < 0:
        parser.error("--max-detections must be non-negative")

    return DemoOptions(
        manifest_path=parsed.manifest_path,
        image_path=parsed.image_path,
        target=parsed.target,
        device_id=parsed.device_id,
        score_threshold=parsed.score_threshold,
        max_detections=parsed.max_detections,
    )


def load_model(options: DemoOptions) -> ptaf.CModelFacade:
    """Load the model manifest with optional execution-target override.

    Args:
        options: Validated demo options.

    Returns:
        Loaded object-detection facade.

    Raises:
        RuntimeError: If the manifest is not an object-detection contract.
    """
    model = ptaf.CModelFacade()
    if options.target == "manifest":
        model.LoadModelConfig(str(options.manifest_path))
    else:
        runtime = ptaf.SRuntimeConfig()
        runtime.SetDeviceId(options.device_id)
        runtime.ClearExecutionTargetPriority()
        target = (
            ptaf.EExecutionTarget.cpu
            if options.target == "cpu"
            else ptaf.EExecutionTarget.cuda
        )
        runtime.AddExecutionTarget(target)
        runtime.SetAllowFallback(False)
        model.LoadModelConfigWithRuntimeConfig(str(options.manifest_path), runtime)

    if model.GetRole() != "object_detection":
        raise RuntimeError("YOLO demo requires role=object_detection in the manifest")
    if model.GetNumInputs() != 1 or model.GetNumOutputs() != 1:
        raise RuntimeError("YOLO demo currently requires one model input and one output")
    return model


def prepare_image(model: ptaf.CModelFacade, image_path: Path) -> ptaf.SFloatTensor:
    """Resize an RGB image and invoke shared HWC-to-NCHW preprocessing.

    Args:
        model: Loaded single-input model facade.
        image_path: Input image path.

    Returns:
        Wrapper-owned float32 NCHW tensor.

    Raises:
        RuntimeError: If the model input is not concrete float32 NCHW RGB.
    """
    input_info = model.GetInputInfo(0)
    shape = list(input_info.shape)
    if (
        input_info.dtype != "float32"
        or len(shape) != 4
        or shape[0] != 1
        or shape[1] != 3
        or shape[2] <= 0
        or shape[3] <= 0
    ):
        raise RuntimeError(
            "YOLO demo expects one float32 input with concrete shape [1,3,H,W]"
        )

    height, width = int(shape[2]), int(shape[3])
    with Image.open(image_path) as source_image:
        rgb_image = source_image.convert("RGB").resize(
            (width, height), Image.Resampling.BILINEAR
        )
        hwc_values = np.asarray(rgb_image, dtype=np.uint8).reshape(-1)

    return ptaf.MakeNchwFloatTensorFromHwcFloat(
        input_info.name,
        hwc_values,
        height,
        width,
        3,
        1.0 / 255.0,
        False,
    )


def print_detections(
    model: ptaf.CModelFacade,
    input_tensor: ptaf.SFloatTensor,
    output_tensor: ptaf.SFloatTensor,
    detections: Sequence[ptaf.SDetection2D],
) -> None:
    """Print a stable summary shared with the native and MATLAB demos.

    Args:
        model: Loaded model facade.
        input_tensor: Prepared model input.
        output_tensor: Raw model output.
        detections: Score-sorted pre-NMS detections.
    """
    print(f"role={model.GetRole()}")
    print(f"backend={model.GetBackendDetail()}")
    print(f"input_name={input_tensor.name}")
    print(f"input_shape={list(input_tensor.shape)}")
    print(f"output_name={output_tensor.name}")
    print(f"output_shape={list(output_tensor.shape)}")
    print(f"detections={len(detections)}")
    for index, detection in enumerate(detections):
        print(
            f"detection[{index}]=class={detection.classification.class_id},"
            f"score={detection.classification.score:.6f},"
            f"cx={detection.bounds.center.x:.3f},"
            f"cy={detection.bounds.center.y:.3f},"
            f"w={detection.bounds.size.width:.3f},"
            f"h={detection.bounds.size.height:.3f}"
        )


def run_demo(options: DemoOptions) -> None:
    """Run one facade-backed YOLO inference.

    Args:
        options: Validated demo options.
    """
    if not options.manifest_path.is_file():
        raise FileNotFoundError(f"Missing model manifest: {options.manifest_path}")
    if not options.image_path.is_file():
        raise FileNotFoundError(f"Missing input image: {options.image_path}")

    model = load_model(options)
    input_tensor = prepare_image(model, options.image_path)
    output_tensor = model.InferSingleFloatTensor(input_tensor)

    output_shape = list(output_tensor.shape)
    if not output_shape or output_shape[-1] < 6:
        raise RuntimeError(
            "YOLO demo expects output rows [cx,cy,w,h,objectness,class_scores...]"
        )

    schema = ptaf.SDetectionRowSchema()
    schema.box_encoding = ptaf.EBoundingBoxEncoding.center_xywh
    schema.box_coordinate_0_index = 0
    schema.box_coordinate_1_index = 1
    schema.box_coordinate_2_index = 2
    schema.box_coordinate_3_index = 3
    schema.objectness_index = 4
    schema.first_class_score_index = 5
    schema.class_score_count = output_shape[-1] - 5
    detections = ptaf.DecodeDetectionRows(
        output_tensor,
        schema,
        options.score_threshold,
        options.max_detections,
    )
    print_detections(model, input_tensor, output_tensor, detections)


def main(arguments: Sequence[str] | None = None) -> int:
    """Run the command-line demo and convert runtime errors to exit status.

    Args:
        arguments: Optional argument sequence. Defaults to ``sys.argv``.

    Returns:
        Zero on success, one on runtime failure.
    """
    try:
        run_demo(parse_arguments(arguments))
    except Exception as error:  # Boundary: provide one actionable CLI failure.
        print(f"run_object_detection.py failed: {error}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
